File: subgenerator.py
def subgenerator(start):
  val = start
  while val >= 0:
    yield val
    val -= 1
  
def countdown(start):
  if start >= 0:
    yield start
  yield subgenerator(start - 1)

print(list(countdown(4)))

def countdown_fixed(start):
  if start >= 0:
    yield start
  yield from subgenerator(start - 1) # note the 'yield from' here

print(list(countdown_fixed(4)))

File: subgenerator.js
function* subgenerator(start) {
  var val = start;
  while (val >= 0) {
    yield val;
    val -= 1;
  }
}

function* countdown(start) {
  if (start >= 0) {
    yield start;
  }
  yield subgenerator(start - 1);
}

for (x of countdown(4)) {
  console.log("countdown for/of loop value=", x);
}

// now the same with 'yield from' aka 'yield*'
function* countdown_fixed(start) {
  if (start >= 0) {
    yield start;
  }
  yield* subgenerator(start - 1); // notice the yield*
}

for (x of countdown_fixed(4)) {
  console.log("countdown_fixed for/of loop value=", x);
}

